pkg/costattribution/sample_tracker.go

```diff
@@ -0,0 +1,302 @@
+	st.updateObservations(buf.String(), now, 0, value, &reason)
+}
+
+func (st *SampleTracker) IncrementReceivedSamples(req *mimirpb.WriteRequest, now time.Time) {
+	if st == nil {
+		return
+	}
+
+	// We precompute the cost attribution per request before update Observations and State to avoid frequently update the atomic counters.
+	// This is based on the assumption that usually a single WriteRequest will have samples that belong to the same or few cost attribution groups.
+	dict := make(map[string]int)
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+	for _, ts := range req.Timeseries {
+		st.fillKeyFromLabelAdapters(ts.Labels, buf)
+		dict[string(buf.Bytes())] += len(ts.TimeSeries.Samples) + len(ts.TimeSeries.Histograms)
+	}
+
+	// Update the observations for each label set and update the state per request,
+	// this would be less precised than per sample but it's more efficient
+	var total float64
+	for k, v := range dict {
+		count := float64(v)
+		st.updateObservations(k, now, count, 0, nil)
+		total += count
+	}
+}
+
+func (st *SampleTracker) fillKeyFromLabelAdapters(lbls []mimirpb.LabelAdapter, buf *bytes.Buffer) {
+	buf.Reset()
+	var exists bool
+	for idx, cal := range st.labels {
+		if idx > 0 {
+			buf.WriteRune(sep)
+		}
+		exists = false
+		for _, l := range lbls {
+			if l.Name == cal {
+				exists = true
+				buf.WriteString(l.Value)
+				break
+			}
+		}
+		if !exists {
+			buf.WriteString(missingValue)
+		}
+	}
+}
+
+// updateObservations updates or creates a new observation in the 'observed' map.
+func (st *SampleTracker) updateObservations(key string, ts time.Time, receivedSampleIncrement, discardedSampleIncrement float64, reason *string) {
+	// if not overflow, we need to check if the key exists in the observed map,
+	// if yes, we update the observation, otherwise we create a new observation, and set the overflowSince if the max cardinality is exceeded
+	st.observedMtx.RLock()
+
+	// if overflowSince is set, we only update the overflow counter, this is after the read lock since overflowSince can only be set when holding observedMtx write lock
+	// check it after read lock would make sure that we don't miss any updates
+	if st.overflowSince.Load() > 0 {
+		st.overflowCounter.receivedSample.Add(receivedSampleIncrement)
+		if discardedSampleIncrement > 0 && reason != nil {
+			st.overflowCounter.totalDiscarded.Add(discardedSampleIncrement)
+		}
+		st.observedMtx.RUnlock()
+		return
+	}
+
+	o, known := st.observed[key]
+	if known && st.overflowSince.Load() == 0 {
+		o.lastUpdate.Store(ts.Unix())
+		o.receivedSample.Add(receivedSampleIncrement)
+		if discardedSampleIncrement > 0 && reason != nil {
+			o.discardedSampleMtx.Lock()
+			if _, ok := o.discardedSample[*reason]; ok {
+				o.discardedSample[*reason].Add(discardedSampleIncrement)
+			} else {
+				o.discardedSample[*reason] = atomic.NewFloat64(discardedSampleIncrement)
+			}
+			o.discardedSampleMtx.Unlock()
```

Comment:
Either we switch `discardedSampleMtx` to a RWMutex (and we call `.Add()` while holding just the `RLock`, and we grab the `Lock` for creation, with a second check), or we switch the map to use non-atomic values (because why we need them atomic while we're holding a mutex?)

Given that the amount of reasons is most likely small, I would go for a RWMutex.

If we don't do this, then a customer that is discarding lots of samples would be causing lock contention here.

Reply:
addressed here 5ac64f5df5a2f8442c7a93b3a124d854e8bebb1f